README.md
# net

Helper functions for IP networks.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).



## [Unreleased]

## [0.3.0] 2021-04-22

### Changed

- Improve network sorting function to use the mask.

### Fixed

- Remove duplicate subnets on freeIPRanges.

## [0.2.0] 2020-03-24

### Changed

- Switch from dep to Go modules.
- Use architect orb.



## [0.1.0] 2020-03-19

### Added

- First release.



[Unreleased]: https://github.com/giantswarm/ipam/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/giantswarm/ipam/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/giantswarm/ipam/releases/tag/v0.1.0


range_test.go
package ipam

import (
	"bytes"
	"net"
	"reflect"
	"testing"
)

// ipNetEqual returns true if the given IPNets refer to the same network.
func ipNetEqual(a, b net.IPNet) bool {
	return a.IP.Equal(b.IP) && bytes.Equal(a.Mask, b.Mask)
}

// TestIPToDecimal tests the IPToDecimal function.
func TestIPToDecimal(t *testing.T) {
	tests := []struct {
		ip              string
		expectedDecimal uint32
	}{
		{
			ip:              "10.0.0.0",
			expectedDecimal: 167772160,
		},

		{
			ip:              "10.4.0.0",
			expectedDecimal: 168034304,
		},
		{
			ip:              "255.255.255.255",
			expectedDecimal: 4294967295,
		},
	}

	for index, test := range tests {
		ip := net.ParseIP(test.ip)

		returnedDecimal := IPToDecimal(ip)

		if returnedDecimal != test.expectedDecimal {
			t.Fatalf(
				"%v: unexpected decimal returned.\nexpected: %v, returned: %v",
				index,
				test.expectedDecimal,
				returnedDecimal,
			)
		}
	}
}

// TestDecimalToIP tests the DecimalToIP function.
func TestDecimalToIP(t *testing.T) {
	tests := []struct {
		decimal    uint32
		expectedIP string
	}{
		{
			decimal:    168034304,
			expectedIP: "10.4.0.0",
		},
		{
			decimal:    4294967295,
			expectedIP: "255.255.255.255",
		},
	}

	for index, test := range tests {
		returnedIP := DecimalToIP(test.decimal)

		expectedIP := net.ParseIP(test.expectedIP)
		if !returnedIP.Equal(expectedIP) {
			t.Fatalf(
				"%v: unexpected decimal returned.\nexpected: %v, returned: %v",
				index,
				expectedIP,
				returnedIP,
			)
		}
	}
}

// TestSize tests the Size function.
func TestSize(t *testing.T) {
	tests := []struct {
		mask         int
		expectedSize uint32
	}{
		{
			mask:         23,
			expectedSize: 512,
		},

		{
			mask:         24,
			expectedSize: 256,
		},

		{
			mask:         25,
			expectedSize: 128,
		},
	}

	for index, test := range tests {
		mask := net.CIDRMask(test.mask, 32)

		returnedSize := Size(mask)

		if returnedSize != test.expectedSize {
			t.Fatalf(
				"%v: unexpected size returned.\nexpected: %v, returned: %v",
				index,
				test.expectedSize,
				returnedSize,
			)
		}
	}
}

// TestRange tests the Range function.
func TestRange(t *testing.T) {
	tests := []struct {
		network       string
		expectedStart uint32
		expectedEnd   uint32
	}{
		{
			network:       "10.4.0.0/8",
			expectedStart: 167772160,
			expectedEnd:   184549375,
		},

		{
			network:       "10.4.0.0/16",
			expectedStart: 168034304,
			expectedEnd:   168099839,
		},

		{
			network:       "10.4.0.0/24",
			expectedStart: 168034304,
			expectedEnd:   168034559,
		},

		{
			network:       "172.168.0.0/25",
			expectedStart: 2896691200,
			expectedEnd:   2896691327,
		},
	}

	for index, test := range tests {
		_, network, _ := net.ParseCIDR(test.network)

		start, end := Range(*network)

		if start != test.expectedStart {
			t.Fatalf(
				"%v: unexpected start returned.\nexpected: %v\nreturned: %v\n",
				index,
				test.expectedStart,
				start,
			)
		}

		if end != test.expectedEnd {
			t.Fatalf(
				"%v: unexpected end returned.\nexpected: %v\nreturned: %v\n",
				index,
				test.expectedEnd,
				end,
			)
		}
	}
}

// TestBoundaries tests the Boundaries function.
func TestBoundaries(t *testing.T) {
	tests := []struct {
		network            string
		existing           []string
		expectedBoundaries []uint32
	}{
		// Test an empty set of networks returns the IPs for the
		// start and end of the overall network.
		{
			network:  "10.4.0.0/16",
			existing: []string{},
			expectedBoundaries: []uint32{
				168034304,
				168099839,
			},
		},

		// Test one net at the start of the overall network returns the
		// IPs for the start and end of the overall network,
		// with the smaller network in the middle.
		{
			network:  "10.4.0.0/16",
			existing: []string{"10.4.0.0/24"},
			expectedBoundaries: []uint32{
				168034304,
				168034304,
				168034559,
				168099839,
			},
		},

		// Test a fragmented network, with one subnet.
		{
			network:  "10.4.0.0/16",
			existing: []string{"10.4.1.0/24"},
			expectedBoundaries: []uint32{
				168034304,
				168034560,
				168034815,
				168099839,
			},
		},

		// Test two, different sized, fragment subnets.
		{
			network:  "10.4.0.0/8",
			existing: []string{"10.4.1.0/25", "10.4.9.0/30"},
			expectedBoundaries: []uint32{
				167772160,
				168034560,
				168034687,
				168036608,
				168036611,
				184549375,
			},
		},

		// Test two, different sized, fragment subnets, but with incorrect order.
		{
			network:  "10.4.0.0/8",
			existing: []string{"10.4.9.0/30", "10.4.1.0/25"},
			expectedBoundaries: []uint32{
				167772160,
				168034560,
				168034687,
				168036608,
				168036611,
				184549375,
			},
		},
	}

	for index, test := range tests {
		_, network, _ := net.ParseCIDR(test.network)

		existing := []net.IPNet{}
		for _, e := range test.existing {
			_, n, _ := net.ParseCIDR(e)
			existing = append(existing, *n)
		}

		// TODO: test errors
		returnedBoundaries, _ := Boundaries(*network, existing)

		if !reflect.DeepEqual(returnedBoundaries, test.expectedBoundaries) {
			t.Fatalf(
				"%v: unexpected boundaries returned.\nexpected: %v\nreturned: %v\n",
				index,
				test.expectedBoundaries,
				returnedBoundaries,
			)
		}
	}
}

func TestFreeRanges(t *testing.T) {
	tests := []struct {
		boundaries         []uint32
		expectedFreeRanges []freeRange
	}{
		{
			boundaries: []uint32{168034304, 168099839},
			expectedFreeRanges: []freeRange{
				{
					start: 168034304,
					end:   168099839,
				},
			},
		},

		{
			boundaries: []uint32{
				167772160,
				168034560,
				168034687,
				168036608,
				168036611,
				184549375,
			},
			expectedFreeRanges: []freeRange{
				{start: 167772160, end: 168034560},
				{start: 168034687, end: 168036608},
				{start: 168036611, end: 184549375},
			},
		},
	}

	for index, test := range tests {
		// TODO: test errors
		freeRanges, _ := FreeRanges(test.boundaries)

		if !reflect.DeepEqual(freeRanges, test.expectedFreeRanges) {
			t.Fatalf(
				"%v: unexpected free ranges returned.\nexpected: %v\nreturned: %v\n",
				index,
				test.expectedFreeRanges,
				freeRanges,
			)
		}
	}
}

func TestSpace(t *testing.T) {
	tests := []struct {
		freeRanges           []freeRange
		mask                 int
		expectedIP           uint32
		expectedErrorHandler func(error) bool
	}{
		// Test a case of fitting a network into an unused network.
		{
			freeRanges: []freeRange{
				{start: 168034304, end: 168099839},
			},
			mask:       24,
			expectedIP: 168034304,
		},

		// Test fitting a network into a non-fragmented range.
		{
			freeRanges: []freeRange{
				{start: 168034304, end: 168034304},
				{start: 168034559, end: 168099839},
			},
			mask:       24,
			expectedIP: 168034560,
		},

		// Test adding a network that fills the range
		{
			freeRanges: []freeRange{
				{start: 168034304, end: 168099839}, // 10.4.0.0/16
			},
			mask:       16,
			expectedIP: 168034304,
		},

		// Test adding a network that is too large.
		{
			freeRanges: []freeRange{
				{start: 168034304, end: 168099839}, // 10.4.0.0/16
			},
			mask:                 15,
			expectedErrorHandler: IsSpaceExhausted,
		},
	}

	for index, test := range tests {
		mask := net.CIDRMask(test.mask, 32)

		ip, err := Space(test.freeRanges, mask)

		if err != nil {
			if test.expectedErrorHandler == nil {
				t.Fatalf("%v: unexpected error returned.\nreturned: %v", index, err)
			} else {
				if !test.expectedErrorHandler(err) {
					t.Fatalf("%v: incorrect error returned.\nreturned: %v", index, err)
				}
			}
		} else {
			if test.expectedErrorHandler != nil {
				t.Fatalf("%v: expected error not returned.\nexpected: %v", index, test.expectedErrorHandler)
			}

			if ip != test.expectedIP {
				t.Fatalf(
					"%v: unexpected ip returned. \nexpected: %v\nreturned: %v",
					index,
					test.expectedIP,
					ip,
				)
			}
		}
	}
}

// TestFree tests the Free function.
func TestFree(t *testing.T) {
	tests := []struct {
		network              string
		mask                 int
		existing             []string
		expectedNetwork      string
		expectedErrorHandler func(error) bool
	}{
		// Test that a network with no existing subnets returns the correct subnet.
		{
			network:         "10.4.0.0/16",
			mask:            24,
			existing:        []string{},
			expectedNetwork: "10.4.0.0/24",
		},

		// Test that a network with one existing subnet returns the correct subnet.
		{
			network:         "10.4.0.0/16",
			mask:            24,
			existing:        []string{"10.4.0.0/24"},
			expectedNetwork: "10.4.1.0/24",
		},

		// Test that a network with two existing (non-fragmented) subnets returns the correct subnet.
		{
			network:         "10.4.0.0/16",
			mask:            24,
			existing:        []string{"10.4.0.0/24", "10.4.1.0/24"},
			expectedNetwork: "10.4.2.0/24",
		},

		// Test that a network with an existing subnet, that is fragmented,
		// and can fit one network before, returns the correct subnet.
		{
			network:         "10.4.0.0/16",
			mask:            24,
			existing:        []string{"10.4.1.0/24"},
			expectedNetwork: "10.4.0.0/24",
		},

		// Test that a network with an existing subnet, that is fragmented,
		// and can fit one network before, returns the correct subnet,
		// given a smaller mask.
		{
			network:         "10.4.0.0/16",
			mask:            25,
			existing:        []string{"10.4.1.0/24"},
			expectedNetwork: "10.4.0.0/25",
		},

		// Test that a network with an existing subnet, that is fragmented,
		// but can't fit the requested network size before, returns the correct subnet.
		{
			network:         "10.4.0.0/16",
			mask:            23,
			existing:        []string{"10.4.1.0/24"}, // 10.4.1.0 - 10.4.1.255
			expectedNetwork: "10.4.2.0/23",           // 10.4.2.0 - 10.4.3.255
		},

		// Test that a network with no existing subnets returns the correct subnet,
		// for a mask that does not fall on an octet boundary.
		{
			network:         "10.4.0.0/24",
			mask:            26,
			existing:        []string{},
			expectedNetwork: "10.4.0.0/26",
		},

		// Test that a network with one existing subnet returns the correct subnet,
		// for a mask that does not fall on an octet boundary.
		{
			network:         "10.4.0.0/24",
			mask:            26,
			existing:        []string{"10.4.0.0/26"},
			expectedNetwork: "10.4.0.64/26",
		},

		// Test that a network with two existing fragmented subnets,
		// with a mask that does not fall on an octet boundary, returns the correct subnet.
		{
			network:         "10.4.0.0/24",
			mask:            26,
			existing:        []string{"10.4.0.0/26", "10.4.0.128/26"},
			expectedNetwork: "10.4.0.64/26",
		},

		// Test a setup with multiple, fragmented networks, of different sizes.
		{
			network: "10.4.0.0/24",
			mask:    29,
			existing: []string{
				"10.4.0.0/26",
				"10.4.0.64/28",
				"10.4.0.80/28",
				"10.4.0.112/28",
				"10.4.0.128/26",
			},
			expectedNetwork: "10.4.0.96/29",
		},

		// Test where a network the same size as the main network is requested.
		{
			network:         "10.4.0.0/16",
			mask:            16,
			existing:        []string{},
			expectedNetwork: "10.4.0.0/16",
		},

		// Test a setup where a network larger than the main network is requested.
		{
			network:              "10.4.0.0/16",
			mask:                 15,
			existing:             []string{},
			expectedErrorHandler: IsMaskTooBig,
		},

		// Test where the existing networks are not ordered.
		{
			network:         "10.4.0.0/16",
			mask:            24,
			existing:        []string{"10.4.1.0/24", "10.4.0.0/24"},
			expectedNetwork: "10.4.2.0/24",
		},

		// Test where the existing networks are fragmented, and not ordered.
		{
			network:         "10.4.0.0/16",
			mask:            24,
			existing:        []string{"10.4.2.0/24", "10.4.0.0/24"},
			expectedNetwork: "10.4.1.0/24",
		},

		// Test where the range is full.
		{
			network:              "10.4.0.0/16",
			mask:                 17,
			existing:             []string{"10.4.0.0/17", "10.4.128.0/17"},
			expectedErrorHandler: IsSpaceExhausted,
		},
	}

	for index, test := range tests {
		_, network, _ := net.ParseCIDR(test.network)
		mask := net.CIDRMask(test.mask, 32)

		existing := []net.IPNet{}
		for _, e := range test.existing {
			_, n, _ := net.ParseCIDR(e)
			existing = append(existing, *n)
		}

		returnedNetwork, err := Free(*network, mask, existing)

		if err != nil {
			if test.expectedErrorHandler == nil {
				t.Fatalf("%v: unexpected error returned.\nreturned: %v", index, err)
			} else {
				if !test.expectedErrorHandler(err) {
					t.Fatalf("%v: incorrect error returned.\nreturned: %v", index, err)
				}
			}
		} else {
			_, expected, _ := net.ParseCIDR(test.expectedNetwork)
			if !ipNetEqual(returnedNetwork, *expected) {
				t.Fatalf(
					"%v: unexpected network returned. \nexpected: %s (%#v, %#v) \nreturned: %s (%#v, %#v)",
					index,

					expected.String(),
					expected.IP,
					expected.Mask,

					returnedNetwork.String(),
					returnedNetwork.IP,
					returnedNetwork.Mask,
				)
			}
		}
	}
}


error.go
package ipam

import (
	"github.com/giantswarm/microerror"
)

var maskTooBigError = microerror.New("mask too big")

// IsMaskTooBig asserts maskTooBigError.
func IsMaskTooBig(err error) bool {
	return microerror.Cause(err) == maskTooBigError
}

var nilIPError = microerror.New("nil IP")

// IsNilIP asserts nilIPError.
func IsNilIP(err error) bool {
	return microerror.Cause(err) == nilIPError
}

var ipNotContainedError = microerror.New("IP not contained")

// IsIPNotContained asserts ipNotContainedError.
func IsIPNotContained(err error) bool {
	return microerror.Cause(err) == ipNotContainedError
}

var maskIncorrectSizeError = microerror.New("mask incorrect size")

// IsMaskIncorrectSize asserts maskIncorrectSizeError.
func IsMaskIncorrectSize(err error) bool {
	return microerror.Cause(err) == maskIncorrectSizeError
}

var spaceExhaustedError = microerror.New("space exhausted")

// IsSpaceExhausted asserts spaceExhaustedError.
func IsSpaceExhausted(err error) bool {
	return microerror.Cause(err) == spaceExhaustedError
}

var incorrectNumberOfBoundariesError = microerror.New("incorrect number of boundaries")

// IsIncorrectNumberOfBoundaries asserts incorrectNumberOfBoundariesError.
func IsIncorrectNumberOfBoundaries(err error) bool {
	return microerror.Cause(err) == incorrectNumberOfBoundariesError
}

var incorrectNumberOfFreeRangesError = microerror.New("incorrect number of free ranges")

// IsIncorrectNumberOfFreeRangesError asserts incorrectNumberOfFreeRangesError.
func IsIncorrectNumberOfFreeRanges(err error) bool {
	return microerror.Cause(err) == incorrectNumberOfFreeRangesError
}


range.go
package ipam

import (
	"bytes"
	"encoding/binary"
	"math"
	"net"
	"sort"

	"github.com/giantswarm/microerror"
)

// IPToDecimal converts a net.IP to a uint32.
func IPToDecimal(ip net.IP) uint32 {
	t := ip
	if len(ip) == 16 {
		t = ip[12:16]
	}

	return binary.BigEndian.Uint32(t)
}

// DecimalToIP converts a uint32 to a net.IP.
func DecimalToIP(ip uint32) net.IP {
	t := make(net.IP, 4)
	binary.BigEndian.PutUint32(t, ip)

	return t
}

// Size takes a mask, and returns the number of addresses.
func Size(mask net.IPMask) uint32 {
	ones, _ := mask.Size()
	size := uint32(math.Pow(2, float64(32-ones)))

	return size
}

// Range takes an IPNet, and returns the start and end IPs.
func Range(network net.IPNet) (uint32, uint32) {
	start := IPToDecimal(network.IP)
	end := start + Size(network.Mask) - 1

	return start, end
}

// Boundaries takes a network, and a list of networks. A list of the IP addresses,
// as uint32 is returned. The start and end of the first supplied networks are the
// first and last entries, with each networks first and last IPs between.
func Boundaries(network net.IPNet, existing []net.IPNet) ([]uint32, error) {
	sort.Sort(IPNets(existing))

	rangeStart, rangeEnd := Range(network)

	boundaries := []uint32{rangeStart}

	for _, existingNetwork := range existing {
		start, end := Range(existingNetwork)
		boundaries = append(boundaries, start, end)
	}

	boundaries = append(boundaries, rangeEnd)

	// Invariant: The number of boundaries must be even.
	if len(boundaries)%2 != 0 {
		return nil, microerror.Maskf(
			incorrectNumberOfBoundariesError, "%v ipnets, %v boundaries", len(existing), len(boundaries),
		)
	}

	return boundaries, nil
}

// FreeRanges converts a list of boundaries into a list of free ranges.
// e.g: [168034304, 168099839] to [{168034304, 168099839}].
// These free ranges are then used to find a free space for a new network.
func FreeRanges(boundaries []uint32) ([]freeRange, error) {
	// Invariant: The number of boundaries must be even.
	if len(boundaries)%2 != 0 {
		return nil, microerror.Maskf(
			incorrectNumberOfBoundariesError, "%v boundaries", len(boundaries),
		)
	}

	freeRanges := []freeRange{}

	for i := 0; i < len(boundaries)-1; i = i + 2 {
		freeRanges = append(
			freeRanges,
			freeRange{start: boundaries[i], end: boundaries[i+1]},
		)
	}

	// Invariant: Number of free ranges is half the number of boundaries.
	if len(freeRanges) != len(boundaries)/2 {
		return nil, microerror.Maskf(
			incorrectNumberOfFreeRangesError, "%v boundaries, %v free ranges", len(boundaries), len(freeRanges),
		)
	}

	return freeRanges, nil
}

// Space takes a list of free ranges, and returns the first IP that has space.
func Space(freeRanges []freeRange, mask net.IPMask) (uint32, error) {
	for i := 0; i < len(freeRanges); i++ {
		free := freeRanges[i]

		if free.end-free.start+1 >= Size(mask) {
			start := free.start

			// In the case that we are not at the start of the overall range,
			// we need to increment to the start of the next range.
			if i != 0 {
				start++
			}

			return start, nil
		}
	}

	return 0, microerror.Maskf(spaceExhaustedError, "tried to fit: %v", mask)
}

// Free takes a network, a mask, and a list of networks.
// An available network, within the first network, is returned.
func Free(network net.IPNet, mask net.IPMask, existing []net.IPNet) (net.IPNet, error) {
	if Size(network.Mask) < Size(mask) {
		return net.IPNet{}, microerror.Maskf(
			maskTooBigError, "have: %v, requested: %v", network.Mask, mask,
		)
	}

	sort.Sort(IPNets(existing))

	// Get the start and end of each network currently in use.
	boundaries, err := Boundaries(network, existing)
	if err != nil {
		return net.IPNet{}, microerror.Mask(err)
	}

	// Determine the amount of free space between each network in use,
	// as well as betweeen the start and end of the total space.
	freeRanges, err := FreeRanges(boundaries)
	if err != nil {
		return net.IPNet{}, microerror.Mask(err)
	}

	// Attempt to find a free space, of the required size.
	free, err := Space(freeRanges, mask)
	if err != nil {
		return net.IPNet{}, microerror.Mask(err)
	}

	freeIP := DecimalToIP(free)
	freeNetwork := net.IPNet{IP: freeIP, Mask: mask}

	// Invariant: The IP of the network returned should not be nil.
	if freeNetwork.IP == nil {
		return net.IPNet{}, microerror.Mask(nilIPError)
	}

	// Invariant: The IP of the network returned should be contained
	// within the network supplied.
	if !network.Contains(freeNetwork.IP) {
		return net.IPNet{}, microerror.Maskf(
			ipNotContainedError, "%v is not contained by %v", freeNetwork.IP, network,
		)
	}

	// Invariant: The mask of the network returned should be equal to
	// the mask supplied as an argument.
	if !bytes.Equal(mask, freeNetwork.Mask) {
		return net.IPNet{}, microerror.Maskf(
			maskIncorrectSizeError, "have: %v, requested: %v", freeNetwork.Mask, mask,
		)
	}

	return freeNetwork, nil
}


ipam.go
// ipam provides IP address management functionality.
package ipam

import (
	"bytes"
	"encoding/binary"
	"math"
	"math/bits"
	"net"
	"reflect"
	"sort"

	"github.com/giantswarm/microerror"
)

// CalculateSubnetMask calculates new subnet mask to accommodate n subnets.
func CalculateSubnetMask(networkMask net.IPMask, n uint) (net.IPMask, error) {
	if n == 0 {
		return nil, microerror.Maskf(invalidParameterError, "divide by zero")
	}

	// Calculate amount of bits needed to accommodate at least N subnets.
	subnetBitsNeeded := bits.Len(n - 1)

	maskOnes, maskBits := networkMask.Size()
	if subnetBitsNeeded > maskBits-maskOnes {
		return nil, microerror.Maskf(invalidParameterError, "no room in network mask %s to accommodate %d subnets", networkMask.String(), n)
	}

	return net.CIDRMask(maskOnes+subnetBitsNeeded, maskBits), nil
}

// CanonicalizeSubnets iterates over subnets and returns deduplicated list of
// networks that belong to networkRange. Subnets that overlap each other but
// aren't exactly the same are not removed. Subnets are returned in the same
// order as they appear in input.
//
// Example:
//	  networkRange: 192.168.2.0/24
//	  subnets: [172.168.2.0/25, 192.168.2.0/25, 192.168.3.128/25, 192.168.2.0/25, 192.168.2.128/25]
//	  returned: [192.168.2.0/25, 192.168.2.128/25]
//
// Example 2:
//	  networkRange: 10.0.0.0/8
//	  subnets: [10.1.0.0/16, 10.1.0.0/24, 10.1.1.0/24]
//	  returned: [10.1.0.0/16, 10.1.0.0/24, 10.1.1.0/24]
//
func CanonicalizeSubnets(networkRange net.IPNet, subnets []net.IPNet) []net.IPNet {
	// Naive deduplication as net.IPNet cannot be used as key for map. This
	// should be ok for current foreseeable future.
	for i := 0; i < len(subnets); i++ {
		// Remove subnets that don't belong to our desired network.
		if !networkRange.Contains(subnets[i].IP) {
			subnets = append(subnets[:i], subnets[i+1:]...)
			i--
			continue
		}

		// Remove duplicates.
		for j := i + 1; j < len(subnets); j++ {
			if reflect.DeepEqual(subnets[i], subnets[j]) {
				subnets = append(subnets[:j], subnets[j+1:]...)
				j--
			}
		}
	}

	return subnets
}

// Contains returns true when the subnet is a part of the network, false
// otherwise.
func Contains(network, subnet net.IPNet) bool {
	subnetRange := newIPRange(subnet)
	return network.Contains(subnetRange.start) && network.Contains(subnetRange.end)
}

// Free takes a network, a mask, and a list of subnets.
// An available network, within the first network, is returned.
func Free(network net.IPNet, mask net.IPMask, subnets []net.IPNet) (net.IPNet, error) {
	if size(network.Mask) < size(mask) {
		return net.IPNet{}, microerror.Maskf(
			maskTooBigError, "have: %v, requested: %v", network.Mask, mask,
		)
	}

	for _, subnet := range subnets {
		if !network.Contains(subnet.IP) {
			return net.IPNet{}, microerror.Maskf(
				ipNotContainedError, "%v is not contained by %v", subnet.IP, network,
			)
		}
	}

	sort.Sort(ipNets(subnets))

	// Find all the free IP ranges.
	freeIPRanges, err := freeIPRanges(network, subnets)
	if err != nil {
		return net.IPNet{}, microerror.Mask(err)
	}

	// Attempt to find a free space, of the required size.
	freeIP, err := space(freeIPRanges, mask)
	if err != nil {
		return net.IPNet{}, microerror.Mask(err)
	}

	// Invariant: The IP of the network returned should not be nil.
	if freeIP == nil {
		return net.IPNet{}, microerror.Mask(nilIPError)
	}

	freeNetwork := net.IPNet{IP: freeIP, Mask: mask}

	// Invariant: The IP of the network returned should be contained
	// within the network supplied.
	if !network.Contains(freeNetwork.IP) {
		return net.IPNet{}, microerror.Maskf(
			ipNotContainedError, "%v is not contained by %v", freeNetwork.IP, network,
		)
	}

	// Invariant: The mask of the network returned should be equal to
	// the mask supplied as an argument.
	if !bytes.Equal(mask, freeNetwork.Mask) {
		return net.IPNet{}, microerror.Maskf(
			maskIncorrectSizeError, "have: %v, requested: %v", freeNetwork.Mask, mask,
		)
	}

	return freeNetwork, nil
}

// Half takes a network and returns two subnets which split the network in
// half.
func Half(network net.IPNet) (first, second net.IPNet, err error) {
	ones, bits := network.Mask.Size()
	if ones == bits {
		return net.IPNet{}, net.IPNet{}, microerror.Maskf(maskTooBigError, "single IP mask %q is not allowed", network.Mask.String())
	}

	// Bit shift is dividing by 2.
	ones++
	mask := net.CIDRMask(ones, bits)

	// Compute first half.
	first, err = Free(network, mask, nil)
	if err != nil {
		return net.IPNet{}, net.IPNet{}, microerror.Mask(err)
	}

	// Second half is computed by getting next free.
	second, err = Free(network, mask, []net.IPNet{first})
	if err != nil {
		return net.IPNet{}, net.IPNet{}, microerror.Mask(err)
	}

	return first, second, nil
}

// Split returns n subnets from network.
func Split(network net.IPNet, n uint) ([]net.IPNet, error) {
	mask, err := CalculateSubnetMask(network.Mask, n)
	if err != nil {
		return nil, microerror.Mask(err)
	}

	var subnets []net.IPNet
	for i := uint(0); i < n; i++ {
		subnet, err := Free(network, mask, subnets)
		if err != nil {
			return nil, microerror.Mask(err)
		}

		subnets = append(subnets, subnet)
	}

	return subnets, nil
}

// add increments the given IP by the number.
// e.g: add(10.0.4.0, 1) -> 10.0.4.1.
// Negative values are allowed for decrementing.
func add(ip net.IP, number int) net.IP {
	return decimalToIP(ipToDecimal(ip) + number)
}

// decimalToIP converts an int to a net.IP.
func decimalToIP(ip int) net.IP {
	t := make(net.IP, 4)
	binary.BigEndian.PutUint32(t, uint32(ip))

	return t
}

// removeDuplicateSubnets takes a list of subnets.
// It will remove subnets that are included in other CIDRs in the list
func removeDuplicateSubnets(subnets []net.IPNet) []net.IPNet {

	var uniqueSubnets []net.IPNet

	for _, currentSubnet := range subnets {
		var isSubnetIncluded bool

		// Check if current subnet is a subset of any other subnet in the list
		for _, othersubnet := range subnets {
			if !reflect.DeepEqual(currentSubnet, othersubnet) && Contains(othersubnet, currentSubnet) {
				isSubnetIncluded = true
				break
			}
		}

		//If the subnet is not a subset, we add it to the output list
		if !isSubnetIncluded {
			uniqueSubnets = append(uniqueSubnets, currentSubnet)
		}
	}

	return uniqueSubnets
}

// freeIPRanges takes a network, and a list of subnets.
// It calculates available IPRanges, within the original network.
func freeIPRanges(network net.IPNet, subnets []net.IPNet) ([]ipRange, error) {
	freeSubnets := []ipRange{}
	networkRange := newIPRange(network)

	subnets = removeDuplicateSubnets(subnets)

	if len(subnets) == 0 {
		freeSubnets = append(freeSubnets, networkRange)
		return freeSubnets, nil
	}

	{
		// Check space between start of network and first subnet.
		firstSubnetRange := newIPRange(subnets[0])

		// Check the first subnet doesn't start at the start of the network.
		if !networkRange.start.Equal(firstSubnetRange.start) {
			// It doesn't, so we have a free range between the start
			// of the network, and the start of the first subnet.
			end := add(firstSubnetRange.start, -1)
			freeSubnets = append(freeSubnets,
				ipRange{start: networkRange.start, end: end},
			)
		}
	}

	{
		// Check space between each subnet.
		for i := 0; i < len(subnets)-1; i++ {
			currentSubnetRange := newIPRange(subnets[i])
			nextSubnetRange := newIPRange(subnets[i+1])

			// If the two subnets are not contiguous,
			if ipToDecimal(currentSubnetRange.end)+1 != ipToDecimal(nextSubnetRange.start) {
				// Then there is a free range between them.
				start := add(currentSubnetRange.end, 1)
				end := add(nextSubnetRange.start, -1)
				freeSubnets = append(freeSubnets, ipRange{start: start, end: end})
			}
		}
	}

	{
		// Check space between last subnet and end of network.
		lastSubnetRange := newIPRange(subnets[len(subnets)-1])

		// Check the last subnet doesn't end at the end of the network.
		if !lastSubnetRange.end.Equal(networkRange.end) {
			// It doesn't, so we have a free range between the end of the
			// last subnet, and the end of the network.
			start := add(lastSubnetRange.end, 1)
			freeSubnets = append(freeSubnets,
				ipRange{start: start, end: networkRange.end},
			)
		}
	}

	return freeSubnets, nil
}

// ipToDecimal converts a net.IP to an int.
func ipToDecimal(ip net.IP) int {
	t := ip
	if len(ip) == 16 {
		t = ip[12:16]
	}

	return int(binary.BigEndian.Uint32(t))
}

// newIPRange takes an IPNet, and returns the ipRange of the network.
func newIPRange(network net.IPNet) ipRange {
	start := network.IP
	end := add(network.IP, size(network.Mask)-1)

	return ipRange{start: start, end: end}
}

// size takes a mask, and returns the number of addresses.
func size(mask net.IPMask) int {
	ones, _ := mask.Size()
	size := int(math.Pow(2, float64(32-ones)))

	return size
}

// space takes a list of free ip ranges, and a mask,
// and returns the start IP of the first range that could fit the mask.
func space(freeIPRanges []ipRange, mask net.IPMask) (net.IP, error) {
	for _, freeIPRange := range freeIPRanges {
		start := ipToDecimal(freeIPRange.start)
		end := ipToDecimal(freeIPRange.end)

		// When subnet allocations contain various different subnet sizes, it can be
		// that free IP range starts from smaller network than what we are finding
		// for. Therefore we must first adjust the start IP such that it can hold the
		// whole network that we are looking space for.
		//
		// Example: Free IP range starts at 10.1.2.192 and ends 10.1.255.255.
		//          We look for next available /24 network so first suitable
		//          start IP for this would be 10.1.3.0.
		//
		ones, _ := mask.Size()
		trailingZeros := bits.TrailingZeros32(uint32(start))
		for (start < end) && (ones < (32 - trailingZeros)) {
			var mask uint32
			for i := 0; i < trailingZeros; i++ {
				mask |= 1 << uint32(i)
			}

			start = int(uint32(start) | mask)
			start++
			trailingZeros = bits.TrailingZeros32(uint32(start))
		}

		if end-start+1 >= size(mask) {
			return decimalToIP(start), nil
		}
	}

	return nil, microerror.Maskf(spaceExhaustedError, "tried to fit: %v", mask)
}
